opengovsg/plumber
# PLU-125: [EXCEL-7] Usage tracking & rate limiting

## Problem
We need to report our M365 API usage and obey their rate limits.

## Solution
We will do this by adding 2 `beforeRequest` callbacks to log usage and apply rate limits.

### Side note: `M365RetryStepError`
Please ignore this bit. I implemented this because I was iffy about including app-specific logic in our usual exponential backoff handler, but i have a better way around that in a later PR.

But as part of this, I also added a new `delayMs` field to our `ActionRunResult` type, which allows the current action to delay the next step. This enables our delay actions to be generic, so I left it in for review.

## Tests
- Check that M365 API calls are logged
- Set rate limit to 1 and check that M365 calls get rate limited.
- Regression test: check that delay still works.
